blog/_posts/2021-03-25-kubernetes-on-ci.md

```diff
@@ -0,0 +1,150 @@
+
+There were other downsides though:
+
+- When drive-by contributors sent us PRs, the integration tests failed.
+  They didn't have access to to write to the gcr.io bucket. This made me so sad.
+  Contributors felt unwelcome. I never figured out a way to make this secure.
+  
+- We didn't reset the gcr.io bucket between test runs. So it was hard to guarantee
+  that images weren't leaking between tests. For example, if image pushing failed,
+  we wanted to be sure we weren't picking up a cached image from a previous test.
+
+## Strategy #2: Local Registry On a VM
+
+When I revisited this, I wanted to make sure:
+
+- Anyone could write to the image registry.
+
+- The image registry resets between runs.
+
+By this time, `kind` was taking off as the default choice for testing Kubernetes
+itself. `kind` also comes with the ability to run a local registry, so
+you can push images to the registry on `localhost:5000` and pull them from inside `kind`.
+
+I set up a new CI pipeline that:
+ 
+1. Creates a VM.
+
+2. Installs all the dependencies we needed, including Docker.
+
+3. Creates `kind` with a local registry, using their script.
+
+This worked well! And because the registry was local, it was faster too. We
+still use this approach to test `ctlptl` with both `minikube` and `kind`. Here's
+[the CI
+config](https://github.com/tilt-dev/ctlptl/blob/b6f808a09b05b6cf7aa0b3365e4781d2c23e4851/.circleci/config.yml#L30).
+
+But I weren't totally happy! VMs are slower. Upgrading dependencies is a bit
+more heavyweight. We wondered: can we make this work in containers?
+
+## Strategy #3: Local Registry On Remote Docker
+
+The last approach (and the one we use in most of our projects) uses some of the
+tricks that `kubeadm-dind-cluster` uses.
+
+The CI pipeline:
+
+1. Creates a container with our code.
+
+2. Sets up [a remote Docker environment](https://circleci.com/docs/2.0/building-docker-images) outside the container.
+
+3. Starts `kind` with a local registry in the remote Docker.
+
+4. Uses `socat` networking jujitsu to expose the remote registry and Kubernetes
+   cluster inside the local container.
+
+The `socat` makes this a bit tricky. But if you want to fork and hack it, check out
+[this bash script](https://github.com/tilt-dev/kind-local/blob/master/.circleci/with-kind-cluster.sh).
+
+But once it's setup: it's fast, robust, and easy to upgrade dependencies.
+
+## Putting it Together
+
+Hacking together this with bash was the hard part.
+
+I eventually folded the bash logic into [`ctlptl`](https://ctlptl.dev/), which
+is a bit easier for local development. As of `ctlptl` 0.5.0, it will try to detect when
+you have a remote docker environment and set up the `socat` forwarding.
+
+The Go rewrite is _far_ more code, comparing number of lines. But it includes
+error handling, cleanup logic, and idempotency, which makes it more suitable for
+local dev (as opposed to CI environments that you tear down immediately.)
```

Comment:
```suggestion
local dev (as opposed to CI environments that you tear down immediately).
```

but I think you mean "suitable for local dev AS WELL" (not "more suitable for local dev than for CI" which is how this reads). try:

> ...which makes it suitable for local dev as well as CI (i.e. it works well even in environments where you don't just tear down your cluster immediately).

Reply:
done